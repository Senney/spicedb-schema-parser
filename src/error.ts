class LexerError extends Error {
  constructor(line: number, index: number, message: string) {
    super(`${line}:${index}: ${message}`);
  }
}

class SyntaxError extends Error {
  constructor(line: number, index: number, message: string) {
    super(`${line}:${index}: ${message}`);
  }
}

export { LexerError, SyntaxError };
