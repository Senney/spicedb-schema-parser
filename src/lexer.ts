/**
 * Note: Caveats will be implemented later.
 */

import { LexerError } from './error';

export enum TokenType {
  IDENTIFIER,
  DEFINITION,
  RELATION,
  PERMISSION,

  ASSIGN, // =
  EQUAL, // ==
  UNION, // +
  INTERSECTION, // &
  EXCLUDE, // -
  OR, // |
  ARROW, // ->
  MEMBER, // #

  COLON, // :
  LEFT_BRACE, // {
  RIGHT_BRACE, // }
  LEFT_PAREN, // (
  RIGHT_PAREN, // )

  EOF,
}

const spiceDBKeywordLexemes: Record<string, TokenType> = {
  definition: TokenType.DEFINITION,
  relation: TokenType.RELATION,
  permission: TokenType.PERMISSION,
};

export class Token {
  constructor(
    public readonly token: TokenType,
    public readonly lexeme: string,
    public readonly literal: string | null,
    public readonly line: number,
    public readonly col: number,
  ) {}
}

const validIdentifierChars = /[a-z\_][a-z\_\/]*/;

class SpiceDBSchemaLexer {
  private start = 0;
  private current = 0;
  private line = 1;

  constructor(private readonly source: string) {}

  public scan(): Token {
    if (this.isAtEnd()) {
      return new Token(TokenType.EOF, '', null, this.line, this.current);
    }

    const c = this.advance();

    switch (c) {
      case '{':
        return this.emit(this.syntaxToken(TokenType.LEFT_BRACE));
      case '}':
        return this.emit(this.syntaxToken(TokenType.RIGHT_BRACE));
      case '(':
        return this.emit(this.syntaxToken(TokenType.LEFT_PAREN));
      case ')':
        return this.emit(this.syntaxToken(TokenType.RIGHT_PAREN));
      case ':':
        return this.emit(this.syntaxToken(TokenType.COLON));
      case '=':
        if (this.match('=')) {
          return this.emit(this.syntaxToken(TokenType.EQUAL));
        }

        return this.emit(this.syntaxToken(TokenType.ASSIGN));
      case '+':
        return this.emit(this.syntaxToken(TokenType.UNION));
      case '&':
        return this.emit(this.syntaxToken(TokenType.INTERSECTION));
      case '|':
        return this.emit(this.syntaxToken(TokenType.OR));
      case '-':
        if (this.match('>')) {
          return this.emit(this.syntaxToken(TokenType.ARROW));
        }

        return this.emit(this.syntaxToken(TokenType.EXCLUDE));
      case '#':
        return this.emit(this.syntaxToken(TokenType.MEMBER));
      case ' ':
      case '\r':
      case '\t':
        return this.dive();
      case '\n':
        this.line++;
        return this.dive();
      case '/':
        if (this.match('*')) {
          while (
            !(this.advance() === '*' && this.peek() === '/') &&
            !this.isAtEnd()
          ) {
            let a = this.peek();
          }
          this.advance();
          return this.dive();
        }

        if (this.match('/')) {
          while (this.advance() !== '\n' && !this.isAtEnd()) {}
          return this.dive();
        }
      default:
        if (validIdentifierChars.test(c)) {
          return this.emit(this.identifier());
        }
    }

    throw new LexerError(this.line, this.start, `Unrecognized token "${c}".`);
  }

  public scanAll(): Token[] {
    const tokens = [];
    for (
      let token = this.scan();
      token.token !== TokenType.EOF;
      token = this.scan()
    ) {
      tokens.push(token);
    }
    return tokens;
  }

  private isAtEnd(): boolean {
    return this.current >= this.source.length;
  }

  private advance(): string {
    return this.source.charAt(this.current++);
  }

  private peek(): string {
    return this.source.charAt(this.current);
  }

  private match(c: string): boolean {
    if (this.peek() === c) {
      this.advance();

      return true;
    }

    return false;
  }

  private identifier(): Token {
    while (validIdentifierChars.test(this.peek())) {
      this.advance();
    }

    const value = this.source.slice(this.start, this.current);
    const tokenType = spiceDBKeywordLexemes[value] ?? TokenType.IDENTIFIER;

    return new Token(tokenType, value, value, this.line, this.current);
  }

  private emit(token: Token): Token {
    this.start = this.current;

    return token;
  }

  private dive(): Token {
    this.start = this.current;

    return this.scan();
  }

  private syntaxToken(type: TokenType): Token {
    return new Token(
      type,
      this.source.slice(this.start, this.current),
      null,
      this.line,
      this.current,
    );
  }
}

export { SpiceDBSchemaLexer };
