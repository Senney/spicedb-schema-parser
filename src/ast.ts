import { Token } from './lexer';

class Statement {}
class Expression {}

class DefinitionStatement extends Statement {
  constructor(
    public resourceName: string,
    public block: BlockStatement,
  ) {
    super();
  }
}

class BlockStatement extends Statement {
  constructor(public statements: Statement[]) {
    super();
  }
}

class RelationStatement extends Statement {
  constructor(
    public relationName: string,
    public relationDefinition: RelationDefinitionExpression,
  ) {
    super();
  }
}

class PermissionStatement extends Statement {
  constructor(
    public permissionName: string,
    public permissionDefinition: PermissionDefinitionExpression,
  ) {
    super();
  }
}

class RelationDefinitionExpression extends Expression {
  constructor() {
    super();
  }
}

class PermissionDefinitionExpression extends Expression {
  constructor() {
    super();
  }
}

class BinaryExpression extends Expression {
  constructor(
    public lhs: Expression,
    public operator: Token,
    public rhs: Expression,
  ) {
    super();
  }
}

class GroupingExpression extends Expression {
  constructor(public expression: Expression) {
    super();
  }
}

export {
  BinaryExpression,
  BlockStatement,
  DefinitionStatement,
  Expression,
  GroupingExpression,
  PermissionDefinitionExpression,
  PermissionStatement,
  RelationDefinitionExpression,
  RelationStatement,
  Statement,
}
