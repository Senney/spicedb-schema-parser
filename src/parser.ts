import {
  BinaryExpression,
  BlockStatement,
  DefinitionStatement,
  Expression,
  IdentifierExpression,
  MembershipIdentifierExpression,
  PermissionDefinitionExpression,
  PermissionStatement,
  RelationDefinitionExpression,
  RelationStatement,
  Statement,
} from './ast';
import { SyntaxError } from './error';
import { Token, TokenType } from './lexer';

class SpiceDBSchemaParser {
  private current = 0;

  constructor(private tokens: Token[]) {}

  public parse(): Statement[] {
    this.current = 0;

    const statements: Statement[] = [];

    while (this.current < this.tokens.length) {
      statements.push(this.definitionStatement());
    }

    return statements;
  }

  private definitionStatement(): Statement {
    this.consume(TokenType.DEFINITION, 'definition');
    const resourceName = this.consume(TokenType.IDENTIFIER, 'identifier');

    return new DefinitionStatement(resourceName.lexeme, this.blockStatement());
  }

  private blockStatement(): BlockStatement {
    this.consume(TokenType.LEFT_BRACE, '{');

    const statements: Statement[] = [];

    while (!this.check(TokenType.RIGHT_BRACE)) {
      statements.push(this.relationStatement());
    }

    this.consume(TokenType.RIGHT_BRACE, '}');

    return new BlockStatement(statements);
  }

  private relationStatement(): Statement {
    if (!this.check(TokenType.RELATION)) {
      return this.permissionStatement();
    }

    this.consume(TokenType.RELATION, 'relation');
    const identifierToken = this.consume(TokenType.IDENTIFIER, 'identifier');
    this.consume(TokenType.COLON, ':');

    return new RelationStatement(
      identifierToken.lexeme,
      new RelationDefinitionExpression(this.relationDefinitionExpression()),
    );
  }

  private permissionStatement(): Statement {
    this.consume(TokenType.PERMISSION, 'permission');

    return new PermissionStatement(
      this.consume(TokenType.IDENTIFIER, 'identifier').lexeme,
      new PermissionDefinitionExpression(),
    );
  }

  private relationDefinitionExpression(): Expression {
    return this.relationOr();
  }

  private relationOr(): Expression {
    const lhs = this.relationIdentifier();

    if (this.match(TokenType.OR)) {
      const operator = this.previous();
      const rhs = this.relationDefinitionExpression();

      return new BinaryExpression(lhs, operator, rhs);
    }

    return lhs;
  }

  private relationIdentifier(): Expression {
    const symbol = this.consume(TokenType.IDENTIFIER, 'identifier');

    if (this.match(TokenType.MEMBER)) {
      const membership = this.consume(TokenType.IDENTIFIER, 'identifier');

      return new MembershipIdentifierExpression(symbol, membership);
    }

    return new IdentifierExpression(symbol);
  }

  private match(...tokenTypes: TokenType[]): boolean {
    for (const tokenType of tokenTypes) {
      if (this.check(tokenType)) {
        this.advance();
        return true;
      }
    }

    return false;
  }

  private consume(tokenType: TokenType, expected: string): Token {
    if (!this.check(tokenType)) {
      throw new SyntaxError(
        this.peek().line,
        this.peek().col,
        `Expected "${expected}", received "${this.peek().lexeme}"`,
      );
    }

    return this.advance();
  }

  private check(tokenType: TokenType): boolean {
    if (this.isAtEnd()) return false;
    return this.peek().token === tokenType;
  }

  private advance(): Token {
    if (!this.isAtEnd()) this.current++;
    return this.previous();
  }

  private isAtEnd() {
    return this.peek().token === TokenType.EOF;
  }

  private peek(): Token {
    return this.tokens[this.current];
  }

  private previous(): Token {
    return this.tokens[this.current - 1];
  }
}

export { SpiceDBSchemaParser };
