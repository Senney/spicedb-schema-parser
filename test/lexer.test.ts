import { SpiceDBSchemaLexer, Token, TokenType } from '../src/lexer';

describe('SpiceDBSchemaLexer', () => {
  test('should return tokens for a simple schema', () => {
    const schema = `
    /** user represents a user */
    definition user {}
    
    /** group represents a group **/
    definition group {
        /** member is a member of a group, which can be a user or the membership of another group */
        relation member: user | group#member
    }
    
    /** document represents a document */
    definition document {
        /** writer is a writer of the document */
        relation writer: user | group#member
    
        /** reader is a reader of the document */
        relation reader: user | group#member
    
        /** write indicates which user can write to the document */
        permission write = writer
    
        /** read indicates which user can read the document */
        permission read = reader + write

        permission nested = reader->foo & (write + read)
    }
    `;

    const lexer = new SpiceDBSchemaLexer(schema);
    const tokens: Token[] = [];

    for (let token = lexer.scan(); token.token !== TokenType.EOF; token = lexer.scan()) {
      tokens.push(token);
    }

    expect(tokens).toMatchSnapshot();
  });

  test('comment parsing', () => {
    const schema = `
    // Single line comment
    /* Normal block comment */
    /** Semantic 
      * multiline
      */
    /* Block * comment with ** lots of asterisks **/
    `;

    const lexer = new SpiceDBSchemaLexer(schema);
    expect(lexer.scan().token).toEqual(TokenType.EOF);
  });
});
