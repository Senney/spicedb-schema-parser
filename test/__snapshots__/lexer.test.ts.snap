// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpiceDBSchemaLexer > should return tokens for a simple schema 1`] = `
[
  Token {
    "col": 49,
    "lexeme": "definition",
    "line": 3,
    "literal": "definition",
    "token": 1,
  },
  Token {
    "col": 54,
    "lexeme": "user",
    "line": 3,
    "literal": "user",
    "token": 0,
  },
  Token {
    "col": 56,
    "lexeme": "{",
    "line": 3,
    "literal": null,
    "token": 13,
  },
  Token {
    "col": 57,
    "lexeme": "}",
    "line": 3,
    "literal": null,
    "token": 14,
  },
  Token {
    "col": 114,
    "lexeme": "definition",
    "line": 6,
    "literal": "definition",
    "token": 1,
  },
  Token {
    "col": 120,
    "lexeme": "group",
    "line": 6,
    "literal": "group",
    "token": 0,
  },
  Token {
    "col": 122,
    "lexeme": "{",
    "line": 6,
    "literal": null,
    "token": 13,
  },
  Token {
    "col": 240,
    "lexeme": "relation",
    "line": 8,
    "literal": "relation",
    "token": 2,
  },
  Token {
    "col": 247,
    "lexeme": "member",
    "line": 8,
    "literal": "member",
    "token": 0,
  },
  Token {
    "col": 248,
    "lexeme": ":",
    "line": 8,
    "literal": null,
    "token": 12,
  },
  Token {
    "col": 253,
    "lexeme": "user",
    "line": 8,
    "literal": "user",
    "token": 0,
  },
  Token {
    "col": 255,
    "lexeme": "|",
    "line": 8,
    "literal": null,
    "token": 9,
  },
  Token {
    "col": 261,
    "lexeme": "group",
    "line": 8,
    "literal": "group",
    "token": 0,
  },
  Token {
    "col": 262,
    "lexeme": "#",
    "line": 8,
    "literal": null,
    "token": 11,
  },
  Token {
    "col": 268,
    "lexeme": "member",
    "line": 8,
    "literal": "member",
    "token": 0,
  },
  Token {
    "col": 274,
    "lexeme": "}",
    "line": 9,
    "literal": null,
    "token": 14,
  },
  Token {
    "col": 336,
    "lexeme": "definition",
    "line": 12,
    "literal": "definition",
    "token": 1,
  },
  Token {
    "col": 345,
    "lexeme": "document",
    "line": 12,
    "literal": "document",
    "token": 0,
  },
  Token {
    "col": 347,
    "lexeme": "{",
    "line": 12,
    "literal": null,
    "token": 13,
  },
  Token {
    "col": 414,
    "lexeme": "relation",
    "line": 14,
    "literal": "relation",
    "token": 2,
  },
  Token {
    "col": 421,
    "lexeme": "writer",
    "line": 14,
    "literal": "writer",
    "token": 0,
  },
  Token {
    "col": 422,
    "lexeme": ":",
    "line": 14,
    "literal": null,
    "token": 12,
  },
  Token {
    "col": 427,
    "lexeme": "user",
    "line": 14,
    "literal": "user",
    "token": 0,
  },
  Token {
    "col": 429,
    "lexeme": "|",
    "line": 14,
    "literal": null,
    "token": 9,
  },
  Token {
    "col": 435,
    "lexeme": "group",
    "line": 14,
    "literal": "group",
    "token": 0,
  },
  Token {
    "col": 436,
    "lexeme": "#",
    "line": 14,
    "literal": null,
    "token": 11,
  },
  Token {
    "col": 442,
    "lexeme": "member",
    "line": 14,
    "literal": "member",
    "token": 0,
  },
  Token {
    "col": 514,
    "lexeme": "relation",
    "line": 17,
    "literal": "relation",
    "token": 2,
  },
  Token {
    "col": 521,
    "lexeme": "reader",
    "line": 17,
    "literal": "reader",
    "token": 0,
  },
  Token {
    "col": 522,
    "lexeme": ":",
    "line": 17,
    "literal": null,
    "token": 12,
  },
  Token {
    "col": 527,
    "lexeme": "user",
    "line": 17,
    "literal": "user",
    "token": 0,
  },
  Token {
    "col": 529,
    "lexeme": "|",
    "line": 17,
    "literal": null,
    "token": 9,
  },
  Token {
    "col": 535,
    "lexeme": "group",
    "line": 17,
    "literal": "group",
    "token": 0,
  },
  Token {
    "col": 536,
    "lexeme": "#",
    "line": 17,
    "literal": null,
    "token": 11,
  },
  Token {
    "col": 542,
    "lexeme": "member",
    "line": 17,
    "literal": "member",
    "token": 0,
  },
  Token {
    "col": 634,
    "lexeme": "permission",
    "line": 20,
    "literal": "permission",
    "token": 3,
  },
  Token {
    "col": 640,
    "lexeme": "write",
    "line": 20,
    "literal": "write",
    "token": 0,
  },
  Token {
    "col": 642,
    "lexeme": "=",
    "line": 20,
    "literal": null,
    "token": 4,
  },
  Token {
    "col": 649,
    "lexeme": "writer",
    "line": 20,
    "literal": "writer",
    "token": 0,
  },
  Token {
    "col": 736,
    "lexeme": "permission",
    "line": 23,
    "literal": "permission",
    "token": 3,
  },
  Token {
    "col": 741,
    "lexeme": "read",
    "line": 23,
    "literal": "read",
    "token": 0,
  },
  Token {
    "col": 743,
    "lexeme": "=",
    "line": 23,
    "literal": null,
    "token": 4,
  },
  Token {
    "col": 750,
    "lexeme": "reader",
    "line": 23,
    "literal": "reader",
    "token": 0,
  },
  Token {
    "col": 752,
    "lexeme": "+",
    "line": 23,
    "literal": null,
    "token": 6,
  },
  Token {
    "col": 758,
    "lexeme": "write",
    "line": 23,
    "literal": "write",
    "token": 0,
  },
  Token {
    "col": 778,
    "lexeme": "permission",
    "line": 25,
    "literal": "permission",
    "token": 3,
  },
  Token {
    "col": 785,
    "lexeme": "nested",
    "line": 25,
    "literal": "nested",
    "token": 0,
  },
  Token {
    "col": 787,
    "lexeme": "=",
    "line": 25,
    "literal": null,
    "token": 4,
  },
  Token {
    "col": 794,
    "lexeme": "reader",
    "line": 25,
    "literal": "reader",
    "token": 0,
  },
  Token {
    "col": 796,
    "lexeme": "->",
    "line": 25,
    "literal": null,
    "token": 10,
  },
  Token {
    "col": 799,
    "lexeme": "foo",
    "line": 25,
    "literal": "foo",
    "token": 0,
  },
  Token {
    "col": 801,
    "lexeme": "&",
    "line": 25,
    "literal": null,
    "token": 7,
  },
  Token {
    "col": 803,
    "lexeme": "(",
    "line": 25,
    "literal": null,
    "token": 15,
  },
  Token {
    "col": 808,
    "lexeme": "write",
    "line": 25,
    "literal": "write",
    "token": 0,
  },
  Token {
    "col": 810,
    "lexeme": "+",
    "line": 25,
    "literal": null,
    "token": 6,
  },
  Token {
    "col": 815,
    "lexeme": "read",
    "line": 25,
    "literal": "read",
    "token": 0,
  },
  Token {
    "col": 816,
    "lexeme": ")",
    "line": 25,
    "literal": null,
    "token": 16,
  },
  Token {
    "col": 822,
    "lexeme": "}",
    "line": 26,
    "literal": null,
    "token": 14,
  },
]
`;
